agent/grpc/stats_test.go

```diff
@@ -0,0 +1,120 @@
+package grpc
+
+import (
+	"context"
+	"net"
+	"testing"
+	"time"
+
+	"github.com/armon/go-metrics"
+	"github.com/hashicorp/consul/agent/grpc/internal/testservice"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc"
+)
+
+func TestHandler_EmitsStats(t *testing.T) {
+	addr := &net.IPAddr{IP: net.ParseIP("127.0.0.1")}
+	handler := NewHandler(addr)
+
+	simple := &server{name: "the-name"}
+	testservice.RegisterSimpleServer(handler.srv, simple)
+
+	sink := patchGlobalMetrics(t)
+
+	lis, err := net.Listen("tcp", "127.0.0.1:0")
+	require.NoError(t, err)
+
+	go func() {
+		err := handler.srv.Serve(lis)
+		t.Logf("grpc serv: %v", err)
+	}()
+	defer func() {
+		err := handler.Shutdown()
+		t.Logf("grpc server shutdown: %v", err)
+	}()
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	conn, err := grpc.DialContext(ctx, lis.Addr().String(), grpc.WithInsecure())
+	require.NoError(t, err)
+	defer conn.Close()
+
+	client := testservice.NewSimpleClient(conn)
+	fClient, err := client.Flow(ctx, &testservice.Req{Datacenter: "mine"})
+	require.NoError(t, err)
+
+	// Wait for the first event so that we know the stream is sending.
+	_, err = fClient.Recv()
+	require.NoError(t, err)
+
+	expectedCounter := []metricCall{
+		{key: []string{"testing", "grpc", "server", "request"}, val: 1},
+	}
+	require.Equal(t, expectedCounter, sink.incrCounterCalls)
+	expectedGauge := []metricCall{
+		{key: []string{"testing", "grpc", "server", "active_conns"}, val: 1},
+		{key: []string{"testing", "grpc", "server", "active_streams"}, val: 1},
+		// TODO: why is the count reset to 0 before the client receives the second message?
+		{key: []string{"testing", "grpc", "server", "active_streams"}, val: 0},
```

Comment:
I was surprised to see the gauge being decremented to 0 at this point. Maybe the grpc client reads and buffers the stream of messages in a goroutine? I was expecting the stream to remain open until after the second `fClient.Recv()` or until the context is cancelled.

Reply:
Fixed this problem in #8679